kubeflow/tensorboard/gcp.libsonnet

```diff
@@ -0,0 +1,50 @@
+{
+  local k = import "k.libsonnet",
+  local util = import "kubeflow/core/util.libsonnet",
+  local deployment = k.apps.v1beta1.deployment,
+  local params = super.params,
+
+  tbService:: super.tbService,
+
+  local tbDeployment =
+    super.tbDeployment +
```

Comment:
Why do you need super.tbDeployment?

IIUC this gets resolved in tensorboard-gcp.jsonnet when we say

```
local instance = tensorboard.new(env, params) + gcp;
````

So super inside gcp gets resolved to tensorboard.new(env, params).

How is this better then defining mixins for tbDeployment and tfService in gcp.libsonnet

And then explictly adding those mixins to the deployment and service defined in tensorboard.libsonnet? (I think if we did it this way we wouldn't have to use the super keyword).

One thing I notice is that the individual parts e.g. tbDeployment and tbService end up being the sum of the part defined in tensorboard.libsonnet and the mixin defined in tensorboard-gcp.libsonnet because of the use of the super keyword (I presume).  

But why doesn't the field "all" end up having 4 items since the list is defined in tensorboard.libsonnet and tensorboard-gcp.libsonnet and we are adding them? Did you verify that it only has 2 items in the case of the gcp prototype?



I'm fine with what you have here (assuming we don't duplicate 4 items). I honestly don't know what the right style for ksonnet should be, so I'm happy iterating

Can you add some comments at least explaining where super gets resolved?

Reply:
Thanks for the review @jlewi . I'll add notes on how super is resolved.

The improvements are 
1. we only need to define one 'new(env,params)' method (in tensorboard.libsonnet).
2. it's easier to test gcp, aws since these subtypes are defined in gcp.libsonnet, aws.libsonnet
3. any attribute/method defined in the subtype overrides the super attribute/method, including 'all' (i tested this to make sure only 2 where defined)
4. it's matches jsonnet best practices more than the old way of creating an instance of the base type inside the subtype. When referring to super.tbDeployment, it will resolve to any object that has a tbDeployment attribute.